lib/ansible/modules/system/firewalld_ipset.py

```diff
@@ -0,0 +1,267 @@
+    immediate: true
+    addresses:
+      - 10.0.0.1/16
+
+# block the bad sets
+- name: Block inbound traffic from the ipsets we wish to block
+  firewalld:
+    source: "{{ item }}"
+    zone: block
+    state: enabled
+    immediate: true
+    permanent: true
+  with_items:
+    - 'ipset:blocklist'
+    - 'ipset:blocklist-mac'
+
+# Allow from clients we have authorized to the internal zone
+- name: Allow traffic from the allowlist to the internal zone
+  firewalld:
+    source: 'ipset:allowlist'
+    zone: internal
+    state: enabled
+    immediate: true
+    permanent: true
+
+# expose a sensitive service to the authorized clients
+- name: expose port 3000 on internal zone, now authorized clients can reach
+  firewalld:
+    port: 3000/tcp
+    zone: internal
+    state: enabled
+    immediate: true
+    permanent: true
+'''
+
+RETURN = '''
+changed:
+  description: was the state modified after execution
+  returned: always
+  type: bool
+  sample: true
+firewalld_ipset_name:
+  description: The name of the specified ipset
+  returned: always
+  type: str
+  sample: the-authorized-clients
+firewalld_ipset_addresses:
+  description: the list of addresses we are looking to manipulate for an ipset
+  returned: always
+  type: list
+  sample: [ "8.8.8.8", "1.1.1.1", "8.8.7.7" ]
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule
+
+try:
+    from firewall.client import FirewallClient
+    from firewall.client import FirewallClientIPSetSettings
+except ImportError:
+    # not sure of the best way to handle these imports -- maybe importing
+    # FirewallTransaction?
+    pass
+
+
+def firewalld_state(client, indata):
+    if indata['immediate']:
+        client.reload()
+    if indata['permanent']:
+        client.runtimeToPermanent()
+
+
+def run():
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(choices=['present', 'absent'], required=True),
+        settype=dict(choices=[
+            "hash:ip", "hash:ip,port", "hash:ip,port,ip",
+            "hash:ip,port,net", "hash:ip,mark",
+            "hash:net", "hash:net,net",
+            "hash:net,port", "hash:net,port,net", "hash:net,iface",
+            "hash:mac"
+        ], required=True),
+        permanent=dict(type='bool', required=False, default=False),
+        immediate=dict(type='bool', required=False, default=False),
+        addresses=dict(type='list', required=False)
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    # setup the FirewallDClient
+    client = FirewallClient()
+    sets = client.getIPSets()
+    settings = FirewallClientIPSetSettings()
+    settings.setType(module.params['settype'])
+    config = client.config()
+
+    # construct return data
+    result = {
+        "changed": False,
+        "firewalld_ipset_name": module.params['name'],
+        "firewalld_ipset_addresses": module.params['addresses']
+    }
+
+    if module.check_mode:
```

Comment:
Check mode should check system state and return `changed=True` in the event something would have changed if the user were to have executed a task using this module not in check mode.